### ISSUE TITLE
Depth is Incremented on Call from Secondary Collector

### ISSUE THREAD

**Issue by IAmStoxe** [gocolly/colly#522]
First off - This is a great tool. I have been working with it for about 2 weeks now on my tool. I am trying to better understand how I might achieve what I'm looking to accomplish.

Consider the following:

I have two collectors, pageCollector (`pc`) and jsCollector (`jc`)

- Page Collector
  - `pc` handles all non .js file `href`s and sends it back to the same `pc` (This should increase depth as I understand it)
  - `pc` sends all .js files found to `jc`
  - `jc` requests .js file and parsed links out
    - with those parsed links it sends them back to `pc` (This _shouldn't_ increase depth as I understand it)

My problem is this is increasing the depth, but it is (in my thinking) the same step. Meaning if I request a page, and find a JS file, parse a link from it and request it back on `pc` my depth increases twice. I really hope I'm explaining this properly.

Example: https://github.com/IAmStoxe/urlgrab/blob/develop/main.go#L499

Thankfully my code is fairly well commented so I hope it can help provide context:

_(Links only provided for context of my question - I promise)_

**Comment by asciimoo** [comment]
That's interesting.. Depth should only increase if you create new requests with the `Request` object instead of the `Collector` which is not the case in your code. I have to do further tests to verify this bug.

> (Links only provided for context of my question - I promise)

Great project, feel free to add it to https://github.com/gocolly/colly#other-projects-using-colly

**Comment by asciimoo** [comment]
I tested the depth increase and I couldn't reproduce the issue. Added some tests as well, but those also run without error: 66e0188

**Comment by IAmStoxe** [comment]
Would it make any difference that I'm instantiating the collectors independently and not cloning via .Clone() ?

https://github.com/IAmStoxe/urlgrab/blob/develop/main.go#L153

https://github.com/IAmStoxe/urlgrab/blob/develop/main.go#L160

Side note: If .OnRequest fires and I call . Abort () would something have already increased? No right?

I'm racking my brain trying to figure it out. I am going to rework my code to be mode modular a bit more and perhaps in that rework I'll figure out my underlying issue a little better.

I'm not collecting results in a proper golang way; it's not thread safe. So maybe in reworking this I'll solve my other problem?

_P.S. Really, thank you immensely for taking the time to address this and thank you very much for the compliments_

**Comment by asciimoo** [comment]
> Would it make any difference that I'm instantiating the collectors independently and not cloning via .Clone() ?

Not at all, it does not have any effect on depth. Depth is handled by the `Request` object and it is independent from the `Collector`. Try to print `Request.Depth`, perhaps it can help to understand what is happening.
After quickly reading your code, I'd say `Request.Depth` should never be bigger than 1, because you use `Collector.Visit` and `Collector.Request` functions and those don't increment the depth count. Only `Request.Visit`/`Request.Post` increments depth constantly.

> I'm not collecting results in a proper golang way; it's not thread safe. So maybe in reworking this I'll solve my other problem?

What are the symptoms of reaching max depth? Is there any [max depth error](https://pkg.go.dev/github.com/gocolly/colly/v2?tab=doc#pkg-variables)?

> P.S. Really, thank you immensely for taking the time to address this and thank you very much for the compliments

No problem, thanks for the kind words. =]

**Comment by IAmStoxe** [comment]
> What are the symptoms of reaching max depth? Is there any max depth error?

I am getting max depth errors, which I had to implement checks for. It was strange as I didn't expect it to increment. 

I think I will have a better feel and it will be easier to demonstrate (should it still be an issue) when I add the channel for results and clean it up a bit.

I will follow up in short order for sure - Thanks again!

**Comment by asciimoo** [comment]
Great, thanks!

**Comment by IAmStoxe** [comment]
So I think I found it. If I understand it correctly, if I pass a ctx as on https://github.com/IAmStoxe/urlgrab/blob/develop/main.go#L363 it is intended to increase depth - which it seems to be doing for each of the links it finds which creates a problem.

If I understand the flow as it stands:

- `jc` parses X links from the js file and sends them to `pc` via `pageCollectorVisitErr := e.Request.Visit(parsedUrl.String())` 
  - At this point depth is increased +1 for every link found in the JS file as it has to call `e.Request.Visit(parsedUrl.String()` - *Is this correct?*

As soon as I removed my depth checks it seems to hit max depth pretty dang fast and early on.
